### DISCUSSION TITLE
VirtualCharacter movement tips please

### DISCUSSION THREAD

**Discussion by RoboMage** [jrouwe/JoltPhysics#1339] [Q&A] [answered]
Hey I recently got back to doing physics stuff with Jolt, but I'm having a bit of trouble with getting the movement for the VirtualCharacter to feel right.  The primary issue I'm having is with my capsule sticking to objects belong edges.
(Sorry I dont have debug shapes setup yet)
![Screenshot 2024-11-18 154518](https://github.com/user-attachments/assets/8e2f098f-a3ef-4073-af9f-77732f0c27f8)
![Screenshot 2024-11-18 154339](https://github.com/user-attachments/assets/ea8495e4-dab4-40c2-9855-17a64e426204)

While I would like to add in ledge sticking, I dont want to be sticking to the bottom side of the ledge :P
It's hard to tell from the screenshots but in movement you can basically see the shape of the capsule from the way the character rolls off the side. I tried using IsSupported, tried using the ground normals, but neither helped with this specific part. Should I be doing a raycast or something?

I'm sure the code for the tick is going to be much more complicated by the end, but for now I just want to be able to setup a fairly snappy feeling character.

Any help or tips will be greatly appreciated 😃 

Here's an overly simplified version of the code:
Create:
```
    JPH::RefConst<JPH::Shape> shape = JPH::RotatedTranslatedShapeSettings(
        JPH::Vec3(0, half_height + radius, 0),
        JPH::Quat::sIdentity(),
        new JPH::CapsuleShapeSettings(half_height, radius)
    ).Create().Get();

    
    JPH::Ref<JPH::CharacterVirtualSettings> vchar_settings = new JPH::CharacterVirtualSettings();
    vchar_settings->mMaxSlopeAngle = JPH::DegreesToRadians(45.0f);
    vchar_settings->mMaxStrength = 1000.0f;
    vchar_settings->mShape = shape;
    vchar_settings->mBackFaceMode = JPH::EBackFaceMode::CollideWithBackFaces;
    vchar_settings->mCharacterPadding = 0.1f;
    vchar_settings->mPenetrationRecoverySpeed = 1.0f;
    vchar_settings->mPredictiveContactDistance = 0.1f;
    vchar_settings->mSupportingVolume = JPH::Plane(JPH::Vec3::sAxisY(), -radius);
    vchar_settings->mMass = 80.0f;
    vchar_settings->mInnerBodyShape = shape;
    vchar_settings->mInnerBodyLayer = Layers::MOVING;

    JPH::CharacterVirtual* vc = new JPH::CharacterVirtual(
        vchar_settings,
        pos,
        rot,
        &physics_system
    );

    vc->SetListener(new CharacterContactListener);
```

Tick:

        bool is_grounded = false;
        if (vc->IsSupported()) { 
            vec3 ground_normal =  vc->GetGroundNormal();

            // Tried various things using the ground normal, but didn't get anything useful
            float ground_dot = dot(ground_normal, vec3(0,1,0));
            is_grounded = true;
            if ( ground_dot < 1.0) {
                is_grounded = false;
            }
        }

        if (is_grounded) {
            velocity.y = 0.01f;  // Wasn't falling off ledges correctly when setting it to zero
        } else { // Falling
            velocity += gravity * TIME;
            velocity.y = max(velocity.y, -50.0f); // Clamp terminal velocity
        }

        vc->SetLinearVelocity(velocity);

        JPH::CharacterVirtual::ExtendedUpdateSettings update_settings;

        vc->ExtendedUpdate(
            time,
            Physics::GetGravity(),
            update_settings,
            physics_system.GetDefaultBroadPhaseLayerFilter(Layers::MOVING),
            physics_system.GetDefaultLayerFilter(Layers::MOVING),
            {},
            {},
            *temp_allocator
        );

**Comment by jrouwe** [comment]
Are you perhaps using other units than meters?

What type of thing is the character walking on, is it a box shape, or perhaps a mesh shape?

Did you try the updating logic as in the [character test sample](https://github.com/jrouwe/JoltPhysics/blob/232fed57a52e5cd2f0259483892a087cc80507de/Samples/Tests/Character/CharacterVirtualTest.cpp#L128-L163)? 

Your `mCharacterPadding` is rather huge. This can cause artifacts.

Finally you could try `mEnhancedInternalEdgeRemoval = true` on `CharacterBaseSettings`.

Implementing `DebugRendererSimple` should be very easy and may make it more visible what is actually happening.

**Reply by RoboMage** [reply]
> Are you perhaps using other units than meters?

No it should be metric

> What type of thing is the character walking on, is it a box shape, or perhaps a mesh shape?

Just a simple box, current extent is 0.3 on all axis  (The capsule is half_height = 0.5, and radius = 0.25 at the moment)

>Did you try the updating logic as in the [character test sample](https://github.com/jrouwe/JoltPhysics/blob/232fed57a52e5cd2f0259483892a087cc80507de/Samples/Tests/Character/CharacterVirtualTest.cpp#L128-L163)?

Not yet. I'll take a swing at it later tonight if i can

> Your mCharacterPadding is rather huge. This can cause artifacts.

I changed it to 0.001f and that did seem to help a fair bit

> Finally you could try mEnhancedInternalEdgeRemoval = true on CharacterBaseSettings.

No difference in this case

> Implementing DebugRendererSimple should be very easy and may make it more visible what is actually happening.

Just set it up

![jolt_debug_shapes](https://github.com/user-attachments/assets/e6b8d89b-057b-40b7-a04a-e145a2a00101)


Full view (need to work on the depth for the lines)

![Screenshot 2024-11-19 184633](https://github.com/user-attachments/assets/f11100d4-98a8-49f0-a557-6c21a25a289d)

**Reply by jrouwe** [reply, marked as answer]
> I changed it to 0.001f and that did seem to help a fair bit

I would recommend sticking to the default value (don't explicitly initialize it and you'll get the default)
